### PR TITLE
Improve proxy error messages for streams

### PR DESCRIPTION
**Pull request by johnstevenson**
I got my logic wrong, so this needs fixing. The exception messages have also been improved (hopefully).

Previously, this threw a RuntimeException but I changed it to a TransportException to match what I was using for secure-proxy >https not being supported natively.

However, there is also a NoSslException, which looks like it is only specifically used when the application first runs. So what should I be using for no openssl situations?

Also, I could add tests for this, but it seems a bit pointless unless the tests explicity run without openssl. What should I do.

### CODE REVIEW

File: src/Composer/Util/StreamContextFactory.php [hunk 1]
```diff
@@ -75,13 +75,17 @@ public static function initOptions($url, array $options, $forCurl = false)
...
-                    throw new TransportException('You must enable the openssl extension to use a proxy over https.');
+                    throw new TransportException('You must enable the openssl extension to use a secure proxy.');
                 }
-                if (0 === strpos($url, 'https://')) {
-                    throw new TransportException('PHP does not support https requests to a secure proxy.');
+                if ($isHttpsRequest) {
+                    throw new TransportException('You must enable the curl extension to make https requests through a secure proxy.');
                 }
+            } elseif ($isHttpsRequest && !extension_loaded('openssl')) {
+                throw new TransportException('You must enable the openssl extension to make https requests through a proxy.');
```
Comment (Seldaek) [inline src/Composer/Util/StreamContextFactory.php:88]: This code will actually be reached even if no proxy is defined unless I'm mistaken, so maybe it should check before mentioning the proxy in the exception message. I guess https even without proxy will fail later on anyway if openssl is missing.

**Comment by Seldaek** [comment]
Regarding exception classes, IMO it's fine if all the HttpDownloader throws is TransportException, kinda more consistent that way.

File: src/Composer/Util/StreamContextFactory.php (hunk 1 above)
Comment (johnstevenson) [reply src/Composer/Util/StreamContextFactory.php:88]: Ah you're right (and I'm still getting my logic wrong!). I'll change this to only be effective with a proxy (as per the original). Okay?

Comment (Seldaek) [reply src/Composer/Util/StreamContextFactory.php:88]: Sure that sounds fine to me :)

**Comment by johnstevenson** [comment]
>  it's fine if all the HttpDownloader throws is TransportException

Is it worth annotating that the NoSslException exception is only used for the start up then (if that is the intended case)?

**Comment by Seldaek** [comment]
Sure if you want to add that I never argue with more docs :)

**Comment by Seldaek** [comment]
Thanks

**Pull request merged**